File: jarvis_frontend/src/api/vault.ts
export type VaultNote = {
  path: string;
  id: string;
  title: string;
  type: string;
  tags: string[];
  aliases: string[];
  favorite: boolean;
  status?: string | null;
  created?: string | null;
  modified?: string | null;
  links: string[];
  body: string;
  folder: string;
  score?: number;
};

export type VaultMeta = {
  total: number;
  by_type: Record<string, number>;
  root: string;
  favorites: number;
  templates: Record<string, { folder: string; body: string }>;
  folders: string[];
};

export type TagCount = { tag: string; count: number };

export type CreateNoteInput = {
  title: string;
  content?: string;
  folder?: string;
  type?: string;
  tags?: string[];
  status?: string;
  source?: string;
  favorite?: boolean;
};

export type UpdateNoteInput = {
  content?: string;
  title?: string;
  tags?: string[];
  status?: string;
  favorite?: boolean;
};

export type SearchParams = {
  q?: string;
  tag?: string;
  type?: string;
  recent?: boolean;
  favorite?: boolean;
  limit?: number;
};

export function getBaseUrl(): string {
  const server = localStorage.getItem('serverUrl') || 'localhost:8000';
  const trimmed = server.trim().replace(/\/+$/, '');
  return trimmed.startsWith('http') ? trimmed : `http://${trimmed}`;
}

async function request<T>(path: string, init?: RequestInit): Promise<T> {
  const res = await fetch(`${getBaseUrl()}${path}`, {
    headers: { 'Content-Type': 'application/json' },
    ...init,
  });
  if (!res.ok) {
    let detail = res.statusText;
    try {
      const data = await res.json();
      detail = data.detail || detail;
    } catch {
      /* keep statusText */
    }
    throw new Error(detail);
  }
  return (await res.json()) as T;
}

function qs(params: Record<string, string | number | boolean | undefined>): string {
  const search = new URLSearchParams();
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined && value !== '') search.set(key, String(value));
  }
  const str = search.toString();
  return str ? `?${str}` : '';
}

export function getMeta(): Promise<VaultMeta> {
  return request<{ meta: VaultMeta }>('/api/vault/meta').then((d) => d.meta);
}

export function listNotes(params: {
  folder?: string;
  tag?: string;
  type?: string;
  favorite?: boolean;
  limit?: number;
}): Promise<VaultNote[]> {
  return request<{ notes: VaultNote[] }>(`/api/vault/notes${qs(params)}`).then((d) => d.notes);
}

export function getNote(path: string): Promise<VaultNote> {
  const encoded = path.split('/').map(encodeURIComponent).join('/');
  return request<{ note: VaultNote }>(`/api/vault/note/${encoded}`).then((d) => d.note);
}

export function createNote(input: CreateNoteInput): Promise<VaultNote> {
  return request<{ note: VaultNote }>('/api/vault/notes', {
    method: 'POST',
    body: JSON.stringify(input),
  }).then((d) => d.note);
}

export function updateNote(path: string, input: UpdateNoteInput): Promise<VaultNote> {
  const encoded = path.split('/').map(encodeURIComponent).join('/');
  return request<{ note: VaultNote }>(`/api/vault/note/${encoded}`, {
    method: 'PUT',
    body: JSON.stringify(input),
  }).then((d) => d.note);
}

export function deleteNote(path: string): Promise<void> {
  const encoded = path.split('/').map(encodeURIComponent).join('/');
  return request<{ deleted: string }>(`/api/vault/note/${encoded}`, { method: 'DELETE' }).then(
    () => undefined,
  );
}

export function appendNote(path: string, text: string): Promise<VaultNote> {
  return request<{ note: VaultNote }>('/api/vault/note/append', {
    method: 'POST',
    body: JSON.stringify({ path, text }),
  }).then((d) => d.note);
}

export function renameNote(path: string, title: string): Promise<VaultNote> {
  return request<{ note: VaultNote }>('/api/vault/note/rename', {
    method: 'POST',
    body: JSON.stringify({ path, title }),
  }).then((d) => d.note);
}

export function moveNote(path: string, folder: string): Promise<VaultNote> {
  return request<{ note: VaultNote }>('/api/vault/note/move', {
    method: 'POST',
    body: JSON.stringify({ path, folder }),
  }).then((d) => d.note);
}

export function archiveNote(path: string): Promise<VaultNote> {
  return request<{ note: VaultNote }>('/api/vault/note/archive', {
    method: 'POST',
    body: JSON.stringify({ path }),
  }).then((d) => d.note);
}

export function searchNotes(params: SearchParams): Promise<VaultNote[]> {
  return request<{ results: VaultNote[] }>(`/api/vault/search${qs(params)}`).then((d) => d.results);
}

export function getTags(): Promise<TagCount[]> {
  return request<{ tags: TagCount[] }>('/api/vault/tags').then((d) => d.tags);
}

export function getFolders(): Promise<string[]> {
  return request<{ folders: string[] }>('/api/vault/folders').then((d) => d.folders);
}

export function createFolder(folder: string): Promise<{ folder: string }> {
  return request<{ folder: string }>('/api/vault/folders', {
    method: 'POST',
    body: JSON.stringify({ folder }),
  });
}

export function reindexVault(): Promise<{ status: string }> {
  return request<{ status: string }>('/api/vault/index', { method: 'POST' });
}

File: jarvis_backend/routes/vault.py
"""HTTP API for the Markdown memory vault.

All endpoints delegate to the single :class:`VaultManager` singleton
(``routes.state.vault``). Errors are mapped to HTTP status codes:
missing notes -> 404, duplicate paths -> 409, bad input -> 400.
"""

from fastapi import APIRouter, HTTPException, Request

from modules.vault.templates import templates
from routes.state import vault

router = APIRouter(prefix="/api/vault", tags=["vault"])


async def _guard(coro):
    try:
        return await coro
    except FileNotFoundError as e:
        raise HTTPException(status_code=404, detail=f"Note not found: {e}") from e
    except FileExistsError as e:
        raise HTTPException(status_code=409, detail=f"Note already exists: {e}") from e
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e)) from e


@router.get("/meta")
async def vault_meta():
    """Stats, vault root path and templates for the UI."""
    stats = await vault.stats()
    stats["templates"] = templates()
    stats["folders"] = await vault.list_folders()
    return {"meta": stats}


# --- notes ------------------------------------------------------------------


@router.post("/notes")
async def create_note(request: Request):
    body = (
        await request.json()
        if request.headers.get("content-type", "").startswith("application/json")
        else {}
    )
    note = await _guard(
        vault.create_note(
            title=body.get("title", ""),
            folder=body.get("folder"),
            note_type=body.get("type"),
            tags=body.get("tags"),
            content=body.get("content", ""),
            status=body.get("status"),
            source=body.get("source"),
            favorite=body.get("favorite", False),
        )
    )
    return {"note": note}


@router.get("/notes")
async def list_notes(
    folder: str | None = None,
    tag: str | None = None,
    type: str | None = None,
    favorite: bool = False,
    limit: int = 100,
):
    return {
        "notes": await vault.list_notes(
            folder=folder, tag=tag, note_type=type, favorite=favorite, limit=limit
        )
    }


@router.get("/note/{rel_path:path}")
async def get_note(rel_path: str):
    return {"note": await _guard(vault.get_note(rel_path))}


@router.put("/note/{rel_path:path}")
async def update_note(rel_path: str, request: Request):
    body = (
        await request.json()
        if request.headers.get("content-type", "").startswith("application/json")
        else {}
    )
    note = await _guard(
        vault.update_note(
            rel_path,
            content=body.get("content"),
            title=body.get("title"),
            tags=body.get("tags"),
            status=body.get("status"),
            favorite=body.get("favorite"),
        )
    )
    return {"note": note}


@router.delete("/note/{rel_path:path}")
async def delete_note(rel_path: str):
    await _guard(vault.delete_note(rel_path))
    return {"deleted": rel_path}


# --- note actions -----------------------------------------------------------


@router.post("/note/append")
async def append_note(request: Request):
    body = await request.json()
    note = await _guard(vault.append_note(body.get("path", ""), body.get("text", "")))
    return {"note": note}


@router.post("/note/rename")
async def rename_note(request: Request):
    body = await request.json()
    note = await _guard(vault.rename_note(body.get("path", ""), body.get("title", "")))
    return {"note": note}


@router.post("/note/move")
async def move_note(request: Request):
    body = await request.json()
    note = await _guard(vault.move_note(body.get("path", ""), body.get("folder", "")))
    return {"note": note}


@router.post("/note/archive")
async def archive_note(request: Request):
    body = await request.json()
    note = await _guard(vault.archive_note(body.get("path", "")))
    return {"note": note}


@router.get("/links/{rel_path:path}")
async def links(rel_path: str):
    return {"links": await _guard(vault.get_links(rel_path))}


@router.get("/backlinks/{rel_path:path}")
async def backlinks(rel_path: str):
    return {"backlinks": await _guard(vault.get_backlinks(rel_path))}


# --- search / folders / tags ------------------------------------------------


@router.get("/search")
async def search(
    q: str | None = None,
    tag: str | None = None,
    type: str | None = None,
    recent: bool = False,
    favorite: bool = False,
    limit: int = 20,
):
    if not (q or tag or type or favorite or recent):
        raise HTTPException(status_code=400, detail="Provide q, tag, type, favorite or recent")
    return {
        "results": await vault.search(
            query=q, tag=tag, note_type=type, recent=recent, favorite=favorite, limit=limit
        )
    }


@router.get("/folders")
async def folders():
    return {"folders": await vault.list_folders()}


@router.post("/folders")
async def create_folder(request: Request):
    body = await request.json()
    return await _guard(vault.create_folder(body.get("folder", "")))


@router.get("/tags")
async def tags():
    return {"tags": await vault.list_tags()}


@router.post("/index")
async def reindex():
    """Force a full re-sync of the FTS search index from disk."""
    await vault.reindex()
    return {"status": "reindexed"}
